### PR TITLE
[74] implement black (fixes #68 #74)

### LINKED ISSUE #68: Upgrade Python to 3.7.3

### PR DESCRIPTION
**Pull request by w1stler**
`make rebuild`
`make test`

### CODE REVIEW

**Comment by stanislawK** [comment]
Do we want to increase line length (compared to PEP8)?

File: backend/resources/auth.py [hunk 1]
```diff
@@ -32,38 +38,35 @@ def post(self):
...
+            )
 
-        username = result['username']
-        password = result['password']
+        username = result["username"]
+        password = result["password"]
 
         if not (username and password):
-            return {"msg": "Username and password required"},
+            return ({"msg": "Username and password required"},)
```
Comment (magul) [inline backend/resources/auth.py]: That's not a part of black introduction, but here
```
    return ({"msg": "Username and password required"},)
    HTTPStatus.BAD_REQUEST
```
the last line is never reachable and response sent to browser will have status set to `200`

File: Makefile [hunk 2]
```diff
@@ -26,3 +26,6 @@ rebuild: ## Rebuild docker images
 
 test: 	## Run unittests
 	docker-compose exec backend pytest
+
+black: ## Run black
+	docker-compose exec backend black .
```
[the remark this answers is not in the document]
Comment (w1stler) [reply Makefile]: Added.

File: backend/Dockerfile [hunk 3]
```diff
@@ -1,18 +1,26 @@
...
-RUN apt-get update -y && apt-get install -y python3-pip libpq-dev && apt-get autoremove
+RUN apt-get update -y && apt-get install -y libpq-dev && apt-get autoremove
 
 RUN pip3 install --upgrade pip
+RUN apt-get update
+RUN apt-get install -y git
+
+# update pip
+RUN python3.8 -m pip install pip --upgrade
+RUN python3.8 -m pip install wheel
```
[the remark this answers is not in the document]
Comment (w1stler) [reply backend/Dockerfile]: Removed.

File: backend/resources/auth.py (hunk 1 above)
Comment (w1stler) [reply backend/resources/auth.py]: Fixed + removed unused import.

File: Makefile [hunk 4]
```diff
@@ -15,14 +15,20 @@ stop:	## Stop the environment
...
 
-test: 	## Run unittests
+test:	## Run unittests
 	docker-compose exec backend pytest
+
+black:	## Run black
+	docker-compose exec backend black .
+
+black-check:	## Run black
+	docker-compose exec backend black --check .
```
Comment (magul) [inline Makefile]: newline please

Comment (magul) [reply Makefile]: + This command should be added to set of tests on Travis

File: Makefile [hunk 5]
```diff
@@ -15,14 +15,20 @@ stop:	## Stop the environment
...
-psql:   ## Go to the db and make SQL queries
+psql:	## Go to the db and make SQL queries
 	docker-compose exec db psql -U cfp_v3
 
-populate_database: ## Populate database with fake data
+populate_database:	## Populate database with fake data
 	docker-compose exec backend flask populate-database
 
-rebuild: ## Rebuild docker images
+rebuild:	## Rebuild docker images
```
Comment (magul) [inline Makefile]: What is the reason for these changes?

File: backend/Dockerfile [hunk 6]
```diff
@@ -1,18 +1,22 @@
...
 
-RUN apt-get update -y && apt-get install -y python3-pip libpq-dev && apt-get autoremove
+RUN apt-get update -y && apt-get install -y libpq-dev && apt-get autoremove
 
-RUN pip3 install --upgrade pip
+RUN pip install --upgrade pip
+RUN apt-get update
+
+# update pip
+RUN python -m pip install pip --upgrade
```
Comment (magul) [inline backend/Dockerfile]: what's the difference between
```
RUN pip install --upgrade pip
```
and
```
RUN python -m pip install pip --upgrade
```
?

File: backend/Dockerfile [hunk 7]
```diff
@@ -1,18 +1,22 @@
...
 
 ENV LC_ALL=C.UTF-8
 ENV LANG=C.UTF-8
 
-RUN apt-get update -y && apt-get install -y python3-pip libpq-dev && apt-get autoremove
+RUN apt-get update -y && apt-get install -y libpq-dev && apt-get autoremove
 
-RUN pip3 install --upgrade pip
+RUN pip install --upgrade pip
+RUN apt-get update
```
Comment (magul) [inline backend/Dockerfile]: What's the reason for updating apt registry if you're not using it below?

File: backend/Dockerfile [hunk 8]
```diff
@@ -1,18 +1,22 @@
-FROM ubuntu:16.04
+FROM python:3.8
 
 EXPOSE 5000
 
 ENV LC_ALL=C.UTF-8
 ENV LANG=C.UTF-8
```
Comment (magul) [inline backend/Dockerfile]: do we need these lines if we're migrating from `ubuntu:16.04` to `python:3.8`?

File: Makefile (hunk 5 above)
Comment (w1stler) [reply Makefile]: That was helpful when I was doing changes in Dockerfile and wanted to test the changes quickly.

File: backend/Dockerfile (hunk 6 above)
Comment (w1stler) [reply backend/Dockerfile]: As I can see the latter one is recommended for Windows, as there some issues: https://github.com/pypa/pip/issues/1299. I will change it.

File: backend/Dockerfile (hunk 8 above)
Comment (w1stler) [reply backend/Dockerfile]: Removed.

File: backend/Dockerfile (hunk 7 above)
Comment (w1stler) [reply backend/Dockerfile]: Removed.

**Comment by w1stler** [comment]
@magul I think I've responded to all comments. I've also added removing containers command in `make rebuild`. Travis runs `black check`.

File: Makefile (hunk 5 above)
Comment (magul) [reply Makefile]: I'm asking about a switch from spaces to tabs. It introduces inconsistency here (as part of commands still uses spaces).

What do you think about that @arturtamborski ?

Comment (arturtamborski) [reply Makefile]: Absolutely, Makefile is all about the tabs. There are some places where you can indent with spaces but mixing these in Makefile (which heavily depends on tabs) can bite someone in the future.

Also, one idea for ensuring that is to create `.editorconfig` file :)
```
root = true

# Unix-style newlines with a newline ending every file
[*]
end_of_line = lf
insert_final_newline = true

# Set default charset
[*.{js,py}]
charset = utf-8

# 4 space indentation
[*.py]
indent_style = space
indent_size = 4

# Tab indentation (no size specified)
[Makefile]
indent_style = tab

[{package.json,.travis.yml}]
indent_style = space
indent_size = 2
```

File: backend/Dockerfile (hunk 6 above)
Comment (arturtamborski) [reply backend/Dockerfile]: The linked issue is not affecting us there, because we're using Linux-based container.

File: Makefile [hunk 9]
```diff
@@ -1,32 +1,39 @@
...
 	docker-compose build --no-cache
 
-test: 	## Run unittests
+test: ## Run unittests
 	docker-compose exec backend pytest
 
+black: ## Run black
+	docker-compose exec backend black .
+
+black-check: ## Run black
```
Comment (arturtamborski) [inline Makefile]:
```suggestion
black-check: ## Check black
```
